cpp_guidelines/guidelines.md

```diff
@@ -0,0 +1,496 @@
+unnecessarily bloat the screen any further.
+
+**Enforced by:**
+  * `.clang-format`: `AlwaysBreakAfterReturnType: None`
+  * `.clang-format`: `AlwaysBreakBeforeMultilineStrings: false`
+  * `.clang-format`: `Cpp11BracedListStyle: true`
+  * `.clang-format`: `IndentWrappedFunctionNames: false`
+  * `.clang-format`: `KeepEmptyLinesAtTheStartOfBlocks: false`
+  * `.clang-format`: `MaxEmptyLinesToKeep: 2`
+  * `.clang-format`: `SpaceAfterLogicalNot: false`
+  * `.clang-format`: `SpaceBeforeSquareBrackets: false`
+  * `.clang-format`: `SpaceInEmptyBlock: false`
+  * `.clang-format`: `SpaceInEmptyParentheses: false`
+  * `.clang-format`: `SpacesBeforeTrailingComments: 1`
+  * `.clang-format`: `SpacesInAngles: false`
+  * `.clang-format`: `SpacesInConditionalStatement: false`
+  * `.clang-format`: `SpacesInContainerLiterals: false`
+  * `.clang-format`: `SpacesInCStyleCastParentheses: false`
+  * `.clang-format`: `SpacesInParentheses: false`
+  * `.clang-format`: `SpacesInSquareBrackets: false`
+
+**Example:**
+```c++
+int                          // There is no reason for a line break
+foo();
+
+int bar( float x, float y ); // There is no reason for the spaces here
+float array[ 10 ];           // Or here
+baz< double >(...);          // Or here
+```
+
+### Use namespace end comments.
+
+**Reason:** Namespaces can span many lines. It is probably not obvious to
+which namespace a closing `}` belongs.
+
+**Enforced by:**
+  * `.clang-format`: `FixNamespaceComments: true`
+
+**Example:**
+```c++
+namespace bad_example
+{
+}
+
+namespace good_example
+{
+} // good_example
+```
+
+### Put nested namespaces on separate lines.
+
+If available the C++17 syntax is preferred: `namespace first::second`.
+
+**Reason**: Each statement belongs on its own line. This includes namespace
+declarations.
+
+**Enforced by:**
+  * `.clang-format`: `CompactNamespaces: false`
+
+**Example:**
+```c++
+// bad example
+namespace foo { namespace bar {
+}}
+
+// good example
+namespace foo
+{
+    namespace bar
+    {
+    }
+}
+
+// best example
+namespace foo::bar
+-{
+}
+```
+
+### Put constructor initializers and inheritance list members on separate lines.
+
+**Reason:** It is easier to read multiple entries one-by-one than having them
+arbitrarily spread across lines.
+
+**Exception**: All initializers / list members fit into one line.
+
+**Enforced by:**
+  * `.clang-format`: `ConstructorInitializerAllOnOneLineOrOnePerLine: true`
+
+**Example:**
+```c++
+class foo
+{
+    foo()
+        : aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa{val}
+        , bbbbbbbbbbbbb{another}
+        , ccccccccccccccccccccccccccc{value}
+    {}
+};
+```
+
+## Indentation and alignment.
+
+### Use the Allman indentation style.
+
+**Reason:** This gives the code a nice vertical structure and optically
+separates different scopes from another. The K&R indentation style
+unnecessarily increases code density, making it harder to follow its
+structure.
```

Comment:
This contradicts the argument made above to not `overuse whitespace`. Also many IDEs and editors offer nice block scope guides:
![image](https://user-images.githubusercontent.com/1224051/93602027-54ce4f80-f9c2-11ea-81a8-638acba6164a.png)

I would like to hear a stronger rational for Allman style. And I would like vote for a style that keeps the opening brace on the same line ;)

Reply:
I think the given `**Reason**` is correct, but one could elaborate on why it is better at `optically
separat[ing] different scopes`: The K&R example you show would indeed not be improved much by Allman, but K&R breaks down when the block opening statement needs to be broken to the next line:
```c++
if ( something(somethingElse) < moreElse 
    && something(somethingElseElse) > 0) {
    doSomething();
}
```
The Java Variant also breaks down for constructors, which happens more often:
```c++
ClassName::ClassName(Type a, ....)
    : m_a(a), /*more things here so this cannot be inline*/.. {
    m_b = f(a);
}
```
Granted, this could be worked around by having `ContinuationIndentWidth > IndentWidth` and `ConstructorInitializerIndentWidth > IndentWidth`, respectively, but this is not currently proposed, and would also feel strange to me.

Not sure if the
```c++
struct Name {
```
in your example is the Java variant, or either, but this gets into the same trouble when a list of  base classes is added.